### ISSUE TITLE
Support alpine images for aarch64

### ISSUE THREAD

**Issue by mrdziuban** [adoptium/containers#158]
Building the alpine images on an Apple silicon (M1) Mac currently fails with an "Unsupported arch: aarch64" error, e.g. from [this code](https://github.com/adoptium/containers/blob/130db1e5356f5a86868b70809c8311b0aad0c756/17/jre/alpine/Dockerfile.releases.full#L31-L40). Would it be possible to add support for this architecture? Thanks in advance!

**Comment by henrik242** [comment]
@mrdziuban:  As a workaround you can use Bellsoft's images. They already have alpine/musl images for aarch64: https://hub.docker.com/r/bellsoft/liberica-openjdk-alpine-musl/tags?page=1&name=17

**Comment by gounthar** [comment]
I guess it would be more difficult than just adding `aarch64` for `alpine`, am I right? (yes, I did it, and it wasn't enough to generate the right `Dockerfile` in `11/jre/alpine/` for example.

**Comment by karianna** [comment]
> I guess it would be more difficult than just adding `aarch64` for `alpine`, am I right? (yes, I did it, and it wasn't enough to generate the right `Dockerfile` in `11/jre/alpine/` for example. `/8/jdk/ubuntu` is building fine.

Definitely more work required (we haven't looked into it yet)

**Comment by gounthar** [comment]
Let me know if I ever can help.

**Comment by gounthar** [comment]
@karianna : if ever the blocking point was the lack of hardware for building/testing, let me know, and I'll see what I can do.

**Comment by gdams** [comment]
We are getting ever closer with Alpine aarch64 images. We now have some nightly builds available on the [website](https://adoptium.net/temurin/nightly) and aim to get some release builds available soon which can in turn be added to our docker images

**Comment by gounthar** [comment]
Fantastic news @gdams , thanks for the update!

**Comment by setchy** [comment]
@gdams - is there a published roadmap that would show roughly when the nightly aarch64 images will move to the formal release builds?

**Comment by karianna** [comment]
July PSU is our next best bet (no promised though!)

**Comment by gounthar** [comment]
Hi there 👋 

I see regular builds for `aarch64` Alpine, which is nice. 🫂 
Could anyone please give me a link to the July PSU? I don't know how to find the date linked to that event.
Thanks.

**Comment by karianna** [comment]
FYI - The Alpine aarch64 builds aren't very stable at the moment.  The July PSU even hasn't been created yet.

**Comment by gounthar** [comment]
Thanks for the update @sxa and @karianna 🙏 .

**Comment by setchy** [comment]
Hi @sxa, any further updates re timelines for this ticket

**Comment by sxa** [comment]
> Hi @sxa, any further updates re timelines for this ticket

We need to get it to a state where it fully passes the AQA tests. We've got some versions in a reasonable state and are looking to get systems available that we can run the TCK on (We cannot distribute JDKs that do not pass the TCK.

We have seen some hangs with some Alpine/aarch64 builds with some versions though, which will be blockers for releasing those (From memory JDK11 was affected by that).

You can see the latest nightly build jobs, and grab the artifacts to test them if you want, from the following jenkins jobs:

- [jdk8u-alpine-linux-aarch64-temurin](https://ci.adoptopenjdk.net/job/build-scripts/job/jobs/job/jdk8u/job/jdk8u-alpine-linux-aarch64-temurin)
- [jdk11u-alpine-linux-aarch64-temurin](https://ci.adoptopenjdk.net/job/build-scripts/job/jobs/job/jdk11u/job/jdk11u-alpine-linux-aarch64-temurin)
- [jdk17u-alpine-linux-aarch64-temurin](https://ci.adoptopenjdk.net/job/build-scripts/job/jobs/job/jdk17u/job/jdk17u-alpine-linux-aarch64-temurin/)
- [jdk19-linux-aarch64-temurin](https://ci.adoptopenjdk.net/job/build-scripts/job/jobs/job/jdk19/job/jdk19-linux-aarch64-temurin/)

It's also worth pointing out that as an interim measure if you have Rosetta2 installed on your M1 system you can run the x64 version e.g.
```
podman run --rm -it --platform=linux/amd64 eclipse-temurin:17-alpine 
```
Or `s/podman/docker/` if that's what you have installed

**Comment by efenderbosch** [comment]
[Azul Zulu](https://www.azul.com/downloads/?architecture=arm-64-bit&package=jdk) is an alternative vs Bellsoft Liberica for local development:

And [this project](https://hub.docker.com/r/arm64v8/eclipse-temurin) is building arm64 images of Temurin.

**Comment by mv912** [comment]
Is there any updates on java 11 on alpine/aarch64?

**Comment by Devilluminati** [comment]
It would be really good if there was a version for arm64. That would make it so much easier in the daily work of development.

**Comment by paoliniluis** [comment]
Hi @karianna, @sxa, I know that we can’t expect an ETA from this specific issue, but is is being worked on? I would like to know if this work is going to continue to decide between staying in alpine or moving to Debian

**Comment by karianna** [comment]
@paoliniluis We don't have anyone currently working on this (waiting for a volunteer basically :-))

**Comment by paoliniluis** [comment]
Thank you @karianna for your quick reply and honesty

**Comment by OneCricketeer** [comment]
What happens if the aarch64 releases from https://github.com/adoptium/temurin17-binaries/releases
are added to the [alpine image switch case](https://github.com/adoptium/containers/blob/main/17/jdk/alpine/Dockerfile.releases.full#L41-L44)?

**Comment by gounthar** [comment]
For me, it just won't work because it's built for libc and not musl.

**Comment by niemyjski** [comment]
Really need arm support for 20.0.1_9-jdk-alpine, 20-jdk-alpine, 20-alpine
cc https://github.com/carlossg/docker-maven/issues/356

**Comment by xlight05** [comment]
This is a much-needed feature that affects the developer experience. Would be great if we can have this ;(

**Comment by sxa** [comment]
To be clear on what's required here:
- We need to make sure the builds are stable for the versions we're interested in (Ideally we'd start from JDK21, but 17 is a possibility, I've had hangs from JDK11 so that is likely completely out of scope for now). We have already integrated the platform into regular jenkins pipelines - here the are for each version with a quick summary of where it is:
  - [JDK11](https://ci.adoptium.net/job/build-scripts/job/jobs/job/jdk11u/job/jdk11u-alpine-linux-aarch64-temurin) - This was hanging quite regularly so we disabled regular build+test although the last two runs that were done looked ok. I've kicked off a test today of the last tagged release - jdk-11.0.18+7 - to see how that goes.
  - [JDK17](https://ci.adoptium.net/job/build-scripts/job/jobs/job/evaluation/job/jobs/job/jdk17u/job/jdk17u-evaluation-alpine-linux-aarch64-temurin/) - Looks in a reasonable state at the moment :-)
  - [JDK20](https://ci.adoptium.net/job/build-scripts/job/jobs/job/evaluation/job/jobs/job/jdk20u/job/jdk20u-evaluation-alpine-linux-aarch64-temurin/) - Build failure, although that's likely not too hard to fix as it's currently trying to use the wrong boot JDK (JDK18 for JDK20 which isn't supported)
  - [JDK21](https://ci.adoptium.net/job/build-scripts/job/jobs/job/evaluation/job/jobs/job/jdk21/job/jdk21-evaluation-alpine-linux-aarch64-temurin/) - Build failure with a missing class - potentially because it uses a JDK20 for the bootstrapping and that build is quite out of date now with the JDK20 pipelines failing
- Once those are done we need to run the TCK compliance suite before we can ship on this platform (That part of it is restricted access sadly, but if we can get the normal build and AQA test jobs to a state where they are all running cleanly then we can look at doing that side of it)
- Once the builds are clean and passing the TCK we will be able to generate and publish container images to dockerhub alongside the existing ones.
As @karianna said earlier, [volunteers to move this forward](https://github.com/adoptium/containers/issues/158#issuecomment-1484270992) are welcome :-) Ideal requirements would be if you have a system on which you can run builds and test suites and diagnose any problems. Quick samples on how to get started
1. Start up the adoptopenjdk/alpine3_build_image container from dockerhub
2. Follow the steps in [this FAQ entry](https://github.com/adoptium/infrastructure/blob/master/FAQ.md#how-do-i-replicate-a-build-failure) which shows you what to run to build the JDK inside the container
3. Follow [this FAQ entry](https://github.com/adoptium/infrastructure/blob/master/FAQ.md#how-do-i-replicate-a-test-failure) as a quick intro on how to run the test suites. The test jobs in jenkins can hopefully provide an idea of the current state of the jobs so you can target your efforts.

**Comment by chadlwilson** [comment]
If we had the ability to "launch" this support with JDK17 (and JRE17) only due to it being the closest to being "ready" that'd be a huge win in my view.

**Comment by sxa** [comment]
> If we had the ability to "launch" this support with JDK17 (and JRE17) only due to it being the closest to being "ready" that'd be a huge win in my view.

The most likely scenario is to attempt to it with JDK21 - which is LTS and will be GA in September, but that is of course subject to having people to analyse the test results and ensure they are good enough for us to release it.

As an update to what I said in the earlier comment, [JDK20](https://ci.adoptium.net/job/build-scripts/job/jobs/job/evaluation/job/jobs/job/jdk20u/job/jdk20u-evaluation-alpine-linux-aarch64-temurin/) and [JDK21](https://ci.adoptium.net/job/build-scripts/job/jobs/job/evaluation/job/jobs/job/jdk21/job/jdk21-evaluation-alpine-linux-aarch64-temurin/68/console) (which was blocked on JDK20) are now able to build successfully so they are in a reasonable state too, subject to testing.

**Comment by gounthar** [comment]
The binaries have been [working for us](https://github.com/jenkinsci/docker-agent/blob/master/alpine/21/Dockerfile) for a few weeks.
Hopefully, the official images will follow up.

**Comment by smlambert** [comment]
Adding this to the PMC-agenda, as there appears to be some expectations for September, but I presume this is not going to be the case (not only due to lack of machines in temurin-compliance).  Also, it has not to date been added to the 2023 Program Plan (slide 27 pasted below).

<img width="908" alt="Screenshot 2023-09-14 at 11 34 01 AM" src="https://github.com/adoptium/containers/assets/2836948/e527af73-742e-4689-a82f-63d2b021c455">

**Comment by tellison** [comment]
@smlambert FWIW the page you reference is from an earlier version of the project plan. The latest plan update (Feb 2023 update) shows the following table - but no material difference for that platform for the reasons you state.

![Screenshot 2023-09-19 at 11 27 54](https://github.com/adoptium/containers/assets/621161/5456bb35-c05f-4ed7-8ec7-ed2133f9309f)

**Comment by sxa** [comment]
FYI The latest run (73) of https://ci.adoptium.net/job/build-scripts/job/jobs/job/evaluation/job/jobs/job/jdk21/job/jdk21-evaluation-alpine-linux-aarch64-temurin/ looks pretty good (I believe it's only showing some failures we've seen on x64). SO while I'm not making any promises or setting any target dates if that's consistent then we're close to being able to ship JDK21 on this platform if we can complete a TCK run successfully.

**Comment by gounthar** [comment]
That's fantastic news, Stewart!
Thanks. 👍

**Comment by sxa** [comment]
We have done a release build and validated that it passes all of the AQA test suite (Not surprising since it's the same code level as the one I mentioned above!) so the TCK is the only thing outstanding (And we do not yet have the TCK material for 21 so none of the platforms have completed that part yet for JDK21)

**Comment by tellison** [comment]
The PMC discussed this topic at the last meeting. Provided the build passes the applicable compatibility tests we will release it as tertiary (Priority 3) platform. Additional work will be required to deliver as container images etc.

**Comment by ipleten** [comment]
We would appreciate Java 17 for alpine-aarc64. It's a big problem for developers on new Macs M-series.

**Comment by sxa** [comment]
> We would appreciate Java 17 for alpine-aarc64. It's a big problem for developers on new Macs M-series.

Similar to JDK21 there are nightly builds available through the website at https://adoptium.net/en-GB/temurin/nightly/ or via the API, however unlike JDK21 those are latest nightly development snapshots at present instead of builds based on specific tags.

At present we do not have any specific plan to release this version (especially in light of 21 being the new latest LTS) but you're welcome to try it and give feedback on stability etc.

**Comment by gounthar** [comment]
That's excellent news, thanks you!

**Comment by sxa** [comment]
container now live too as `eclipse-temurin:21-alpine` :-) I haven't tested specifically on a mac but I can't think of a reason why it wouldn't work:
```
sxa@test-equinix-ubuntu2004-armv8l-02:~$ uname -a
Linux test-equinix-ubuntu2004-armv8l-02 5.4.0-150-generic #167-Ubuntu SMP Mon May 15 17:36:35 UTC 2023 aarch64 aarch64 aarch64 GNU/Linux
sxa@test-equinix-ubuntu2004-armv8l-02:~$ docker run eclipse-temurin:21-alpine java -version
openjdk version "21" 2023-09-19 LTS
OpenJDK Runtime Environment Temurin-21+35 (build 21+35-LTS)
OpenJDK 64-Bit Server VM Temurin-21+35 (build 21+35-LTS, mixed mode, sharing)
sxa@test-equinix-ubuntu2004-armv8l-02:~$ 
```

**Comment by gdams** [comment]
Closing this issue now. An APK installer for Alpine will be published in the coming days. Thanks to everyone who requested aarch64 builds!

**Comment by gounthar** [comment]
Thank you so much for the amazing work.
We'll be able to propose Jenkins agents for amd64 and aarch64 with jdk21 thanks to you.

**Comment by xlight05** [comment]
Thanks a lot for this guys. Please do consider this for Java 17 also if possible =)

**Comment by sxa** [comment]
> Thank you so much for the amazing work. We'll be able to propose Alpine Jenkins agents for amd64 and aarch64 with jdk21 thanks to you.

We are using a Temurin build on our own jenkins agents (not all yet) so hopefully you won't encounter too many problems with it in that environment @gounthar :-)

**Comment by gounthar** [comment]
All should go fine, thanks for your testimony, Stewart. 👍

**Comment by rtrad89** [comment]
> Thanks a lot for this guys. Please do consider this for Java 17 also if possible =)

Seconded!
